src/convert_case.rs

```diff
@@ -0,0 +1,101 @@
+use crate::kebab::to_kebab;
+use crate::lower_camel::to_lower_camel_case;
+use crate::shouty_kebab::to_shouty_kebab_case;
+use crate::shouty_snake::to_shouty_snake_case;
+use crate::snake::to_snake_case;
+use crate::title::to_title_case;
+use crate::upper_camel::to_upper_camel_case;
+
+/// This trait defines a wrapper function for other case-conversion functions
+/// in that this can tweak the behaviour of those functions based on customization options
+///
+///
+/// ## Example:
+/// ```rust
+///
+/// use heck::{ConvertCase, ConvertCaseOpt, Case};
+///
+/// let sentence = "Aes128";
+/// assert_eq!(sentence.convert_case(ConvertCaseOpt { case: Case::ShoutyKebab, number_starts_word: true }),
+/// "AES-128");
+///
+
+pub trait ConvertCase: ToOwned {
+    /// Convert this type to supported cases with options
+    fn convert_case(&self, opt: ConvertCaseOpt) -> String;
+}
+
+/// Options to tweak how convert_case will behave
+pub struct ConvertCaseOpt {
+    /// supported case
+    pub case: Case,
+    /// whether numbers should start a new word
+    pub number_starts_word: bool,
+}
+
+/// supported cases
+pub enum Case {
+    /// kebab-case
+    Kebab,
+    /// lowerCamelCase
+    LowerCamel,
+    /// SHOUT-KEBAB-CASE
+    ShoutyKebab,
+    /// SHOUTY_SNAKE_CASE
+    ShoutySnake,
+    /// snake_case
+    Snake,
+    /// Title Case
+    Title,
+    /// UpperCamelCase
+    UpperCamel,
+}
+
+pub fn convert_case(s: &str, opt: ConvertCaseOpt) -> String {
```

Comment:
Why is this a separate free function? I think the body of this should just be moved into the trait implementation below.